ztp: tests update for policyGenerator after day2 changes

nishant-parekh opened the pull request:
- reorganized test files splitting constants, getters, assertions & helpers into their own files
- updated test templates per changes made to example templates
- modify policyGen test assertions as per changes made to template
- preserve individual files for generated CRs that can be applied directly to the cluster
- nbde siteconfig test fixes
- alternative implementation for os.Executable for tests

Signed-off-by: nishant-parekh <nparekh@redhat.com>

nishant-parekh commented:
/cc @serngawy 
/cc @imiller0

serngawy commented on ztp/ztp-policy-generator/kustomize/plugin/policyGenerator/v1/policygenerator/policyGen/policyBuilder.go:
```diff
@@ -41,8 +41,15 @@ func (pbuilder *PolicyBuilder) Build(policyGenTemp utils.PolicyGenTemplate) (map
 				// Generate CR without append to policy
-				path := utils.CustomResource + "/" + path
-				policies[path] = resources
+				for _, resource := range resources {
```
I don't think there is a need for loop. if it is multi yaml structure just use the first one. wt do u think ?

nishant-parekh replied:
The reason for this loop was to create a single file per resource instead of one file for all the resources within the template, this section deals with the directly applied CRs, and as discussed we wanted to keep them separate.

nishant-parekh replied:
I think we need to separate this CRs into individual files as they will be applied independently on the cluster using `oc apply <fileName>.yaml` . Based on our discussion in the working session we want to keep this implementation so as to work through the deployment on a file by file basis? @imiller0 what do you think?

nishant-parekh replied:
I have removed the loop from the logic replaced with a size check and processing it as an array of size 1

serngawy commented on ztp/ztp-policy-generator/kustomize/plugin/policyGenerator/v1/policygenerator/policyGenerator.go:
```diff
@@ -15,19 +15,33 @@ var sourcePath string
 var tempPath string
 var outPath string
 var stdout bool
+var isTestRun bool
```
I believe we shouldn't let the test flag come from end user. The test mode should be independent

nishant-parekh replied:
I agree I dont want to add this flag the only issues is the `os.Executable()` call in fileHandler.go that I need to change to a os.getwd() for the test runs. The rest of the changes are just a way to enable that if you have a different suggestion I would be happy to change this approach. 

I considered mocking the fileHandler.go completely for the tests but the policyGenerator doesnt reference a fileHandler as a class variable rather it is encapsulated inside policyBuilder.

serngawy replied:
may consider failing safe, if reading from os.Excutable() fail then read from os.getWd()

nishant-parekh replied:
ohh that sounds reasonable and will work with the actual run as well! I can make that change! thanks for the suggestion! :)

nishant-parekh replied:
updated

serngawy commented on ztp/ztp-policy-generator/kustomize/plugin/policyGenerator/v1/policygenerator/utils/filesHandler.go:
```diff
@@ -11,10 +11,15 @@ type FilesHandler struct {
 	sourceDir string
 	tempDir   string
 	outDir    string
+	isTestRun bool `default:"false"`
```
same, fileHandler tests should be outside of fileHandler

nishant-parekh replied:
I have removed this per our discussion

serngawy reviewed (changes requested):
Better to consider policygen lib  as black box you don't give it a flag to tell I will make Test. You give input and check the expected output.

nishant-parekh commented:
> Better to consider policygen lib as black box you don't give it a flag to tell I will make Test. You give input and check the expected output.

I have changed the approach and added a failsafe within the `readResourceFile` method within `fileHandler.go`

serngawy commented on ztp/ztp-policy-generator/kustomize/plugin/policyGenerator/v1/policygenerator/utils/filesHandler.go:
```diff
@@ -16,7 +16,6 @@ type FilesHandler struct {
 	return &FilesHandler{sourceDir: sourceDir, tempDir: tempDir, outDir: outDir}
 }
-
 func (fHandler *FilesHandler) WriteFile(filePath string, content []byte) error {
```
add line between methods

serngawy reviewed (commented):
small comments

nishant-parekh commented:
/retest
